### PR TITLE
address validation

### PR DESCRIPTION
**Pull request by someanon**

### CODE REVIEW

File: addr/bitcoin.go [hunk 1]
```diff
@@ -0,0 +1,19 @@
+package addr
+
+import (
+	"github.com/bitlum/btcd/chaincfg"
+	"github.com/bitlum/btcutil"
+	"github.com/go-errors/errors"
+)
+
+// validateBitcoin validates bitcoin address according net. Return error if address is invalid.
```
Comment (AndrewSamokhvalov) [inline addr/bitcoin.go]: `Return error if address is invalid`, we have schema for this.

File: addr/bitcoinCash.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+	if net == "regtest" {
+		net = "testnet3"
+	}
+	if toLegacyNet(a.Network) != net {
+		return errors.New("invalid or unsupported net")
+	}
+	return nil
+}
+
+func toLegacyNet(networkType address.NetworkType) string {
```
Comment (AndrewSamokhvalov) [inline addr/bitcoinCash.go]: `netToString`?

File: addr/ethereum.go [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+	eth "github.com/ethereum/go-ethereum/common"
+	"github.com/go-errors/errors"
+)
+
+var re = regexp.MustCompile(fmt.Sprintf(`(?:0x)?[0-9a-fA-F]{%d}`, eth.AddressLength*2))
+
+// validateBitcoin validates ethereum address. Return error if address is invalid.
+func validateEthereum(addr string) error {
+	if !eth.IsHexAddress(addr) || !re.MatchString(addr) {
```
Comment (AndrewSamokhvalov) [inline addr/ethereum.go]: I think there is some function in the ethereum client to validate the address

File: addr/validation.go [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+import (
+	bitcoincfg "github.com/bitlum/btcd/chaincfg"
+	"github.com/bitlum/connector/chains/dash"
+	"github.com/go-errors/errors"
+	litecoincfg "github.com/ltcsuite/ltcd/chaincfg"
+)
+
+// Validate validates asset's addr in specified net. Returns error if address is invalid or asset is unknown.
+// For ETH net is meaningless, you can use "".
+func Validate(asset string, net string, addr string) error {
```
Comment (AndrewSamokhvalov) [inline addr/validation.go]: If I understood correctly than after we imported chains all needed networks we added, and there is no need in this function. Basically we just apply `DecodeAddress` function.

File: addr/validation_test.go [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+
+	type args struct {
+		asset string
+		net   string
+		addr  string
+	}
+	tests := []struct {
+		name    string
+		args    args
+		wantErr bool
```
Comment (AndrewSamokhvalov) [inline addr/validation_test.go]: `inValid` instead?

File: addr/bitcoin.go (hunk 1 above)
Comment (someanon) [reply addr/bitcoin.go]: You mean this commend is superflous?

File: addr/bitcoinCash.go (hunk 2 above)
Comment (someanon) [reply addr/bitcoinCash.go]: ok

File: addr/ethereum.go (hunk 3 above)
Comment (someanon) [reply addr/ethereum.go]: Didn't find

File: addr/validation.go (hunk 4 above)
Comment (someanon) [reply addr/validation.go]: ETH is not added. Litecoin has own validation chaincfg. I thing explicit validation is better when implicit.

File: addr/validation_test.go (hunk 5 above)
Comment (someanon) [reply addr/validation_test.go]: it's generated test

File: addr/bitcoin.go (hunk 1 above)
Comment (AndrewSamokhvalov) [reply addr/bitcoin.go]: yep

File: addr/litecoin.go [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+import (
+	"github.com/go-errors/errors"
+	"github.com/ltcsuite/ltcd/chaincfg"
+	"github.com/ltcsuite/ltcutil"
+)
+
+// validateBitcoin validates litecoin address according net. Return error if address is invalid.
+func validateLitecoin(addr string, net *chaincfg.Params) error {
+	a, err := ltcutil.DecodeAddress(addr, net)
```
Comment (AndrewSamokhvalov) [inline addr/litecoin.go]: a => address

File: addr/validation.go [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+)
+
+// Validate validates asset's addr in specified net. For ETH net is meaningless, you can use "".
+func Validate(asset string, net string, addr string) error {
+	switch asset {
+	case "BTC":
+		var n *bitcoincfg.Params
+		switch net {
+		case "mainnet":
+			n = &bitcoincfg.MainNetParams
```
Comment (AndrewSamokhvalov) [inline addr/validation.go]: n => network

File: addr/validation.go [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+		}
+		return validateLitecoin(addr, n)
+	case "DASH":
+		var n *bitcoincfg.Params
+		switch net {
+		case "mainnet":
+			n = &dash.MainNetParams
+		case "regtest":
+			n = &dash.RegressionNetParams
+		case "testnet3":
```
Comment (AndrewSamokhvalov) [inline addr/validation.go]: testnet3 => testnet

File: addr/validation.go [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+	case "ETH":
+		return validateEthereum(addr)
+	case "LTC":
+		var n *litecoincfg.Params
+		switch net {
+		case "mainnet":
+			n = &litecoincfg.MainNetParams
+		case "regtest":
+			n = &litecoincfg.RegressionNetParams
+		case "testnet4":
```
Comment (AndrewSamokhvalov) [inline addr/validation.go]: testnet4 => testnet

File: addr/validation.go [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+	switch asset {
+	case "BTC":
+		var n *bitcoincfg.Params
+		switch net {
+		case "mainnet":
+			n = &bitcoincfg.MainNetParams
+		case "regtest":
+			n = &bitcoincfg.RegressionNetParams
+		case "testnet3":
+			n = &bitcoincfg.TestNet3Params
```
Comment (AndrewSamokhvalov) [inline addr/validation.go]: use our own configs everywhere

File: addr/validation.go [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+func Validate(asset string, net string, addr string) error {
+	switch asset {
+	case "BTC":
+		var n *bitcoincfg.Params
+		switch net {
+		case "mainnet":
+			n = &bitcoincfg.MainNetParams
+		case "regtest":
+			n = &bitcoincfg.RegressionNetParams
+		case "testnet3":
```
Comment (AndrewSamokhvalov) [inline addr/validation.go]: tesnet3 => testnet

File: addr/validation.go [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+			n = &bitcoincfg.TestNet3Params
+		case "simnet":
+			n = &bitcoincfg.SimNetParams
+		default:
+			return errors.New("invalid or unsupported net")
+		}
+		return validateBitcoin(addr, n)
+	case "BCH":
+		return validateBitcoinCash(addr, net)
```
Comment (AndrewSamokhvalov) [inline addr/validation.go]: Use our configs and add check for special type of bitcoin cash address

Comment (AndrewSamokhvalov) [reply addr/validation.go]: Also add comment here why we have special case for validation

**Pull request merged**